### DISCUSSION TITLE
Environment without tkinter?

### DISCUSSION THREAD

**Discussion by petsuter** [spectrochempy/spectrochempy#484] [Q&A]
Some Python platforms have no tkinter support. Is it possible to avoid this error?
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    import spectrochempy
  File "Python\lib\site-packages\spectrochempy\__init__.py", line 59, in <module>
    from spectrochempy import api
  File "Python\lib\site-packages\spectrochempy\api.py", line 89, in <module>
    from . import core
  File "Python\lib\site-packages\spectrochempy\core\__init__.py", line 366, in <module>
    from tkinter import filedialog
ModuleNotFoundError: No module named 'tkinter'
```
(Qt would be available, but I'm not sure how to enable that and if it avoids tkinter completely. This didn't work:
```
from spectrochempy.application import SpectroChemPy
SpectroChemPy().preferences.use_qt = True
import spectrochempy
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    from spectrochempy.application import SpectroChemPy
  File "Python\lib\site-packages\spectrochempy\__init__.py", line 59, in <module>
    from spectrochempy import api
  File "Python\lib\site-packages\spectrochempy\api.py", line 89, in <module>
    from . import core
  File "Python\lib\site-packages\spectrochempy\core\__init__.py", line 366, in <module>
    from tkinter import filedialog
ModuleNotFoundError: No module named 'tkinter'
```
)

**Comment by atravert** [comment]
Hi,
Thank you for your message. We used tkinter as defaullt because it is light and - in principle - belongs to the standard python library: https://docs.python.org/3/library/index.html .  Could you give us more details on the platform you are using ? 

Best,

**Comment by petsuter** [comment]
At least on Debian Linux tkinter is not available by default (but can be installed) and on MicroPython, the Windows embeddable package or the new WebAssembly platforms like [Pyodide](https://pyodide.org/en/stable/console.html) tkinter is not available at all I think.

**Reply by atravert** [reply]
Hi,
Another user have had recently a similar issue ([here](https://github.com/spectrochempy/spectrochempy/issues/490#issuecomment-1292776307)). 
I have tried to fix this but it is not as easy as for other packages. For instance, as it is normally part of python, iy can't be added in the requirements.txt , etc...
In the mean time, I will add a test to check if tkinter is installed and print an explicit message, e.g.: "tkinter is required by SpectroChemPy and should be installed".

Best,

**Reply by petsuter** [reply]
Thanks for investigating it.

I noticed there's a qt mode. Would it be possible to opt into that somehow as the default? On some platforms it's just not possible to install tkinter as far as I know.

As another example, the matplotlib library also uses tkinter by default, but on platforms without tkinter you can switch to the agg backend instead: 
```
import matplotlib # This line does not require tkinter
matplotlib.use('agg',warn=False, force=True) # Switch from tkinter default backend to agg backend
import matplotlib.pyplot as plt # This line would require tkinter, but now it doesn't.
```
https://matplotlib.org/stable/users/explain/backends.html#the-builtin-backends

If you could somehow allow something like this:
```
import spectrochempy_config
spectrochempy_config.ui_toolkit('qt') # Or even None?
import spectrochempy
```
that would be nice.

Best.

**Reply by fernandezc** [reply]
Hi,
May be I do not understand well what you mean here, but it seems to me that can you do this using what we described here: https://www.spectrochempy.fr/latest/userguide/plotting/plotting.html#Selecting-the-output-window
that is: %matplotlib qt  in Jupiter. 
I am not sure however (as I did not check it yet) that tkinter  is not needed even in this case.

**Reply by petsuter** [reply]
Hi, is it possible to do some equivalent without using Jupyter?

**Reply by fernandezc** [reply]
Let me look into this. I can probably add some configuration flags. For testing, since I can't remove tk from my platform (macOS) without breaking python, I have to install a virtual environment to get Debian first. This may take some time.

**Reply by petsuter** [reply]
FYI you can open a Pyodide shell simply by opening this link in your browser: https://pyodide.org/en/stable/console.html
and try to install spectrochempy:
```
>>> import micropip
>>> await micropip.install('spectrochempy')
```

And here seems to be MicroPython for macOS https://ports.macports.org/port/micropython/

But they have other limitations / differences too.

Maybe just temporarily removing / renaming the tkinter folder in Python's Lib is also an option.

**Comment by fernandezc** [comment]
Thanks for reporting this, and sorry for the very late follow-up.

This issue appears to have been addressed by the current package structure. In recent SpectroChemPy versions, importing spectrochempy no longer imports tkinter.filedialog directly during package initialization; the core package is now lazily loaded. Therefore, environments without tkinter should no longer fail at the simple import spectrochempy step.

Some interactive file-selection features may still require a GUI backend, depending on the platform and usage, but the original import-time failure should no longer occur.

I will mark this discussion as answered/closed for now. Please open a new issue with the current SpectroChemPy version and a minimal reproducer if this still fails on a recent release.
